arangod/VocBase/Methods/UpgradeTasks.cpp

```diff
@@ -155,38 +120,372 @@ bool UpgradeTasks::upgradeGeoIndexes(TRI_vocbase_t& vocbase,
+                                     std::vector<std::shared_ptr<LogicalCollection>>& collections) {
+  if (vocbase.isSystem() && StatisticsFeature::enabled()) {
+    try {
+      ::createIndex(StaticStrings::StatisticsCollection,
+                    arangodb::Index::TRI_IDX_TYPE_SKIPLIST_INDEX, {"time"},
+                    false, false, collections);
+      ::createIndex(StaticStrings::Statistics15Collection,
+                    arangodb::Index::TRI_IDX_TYPE_SKIPLIST_INDEX, {"time"},
+                    false, false, collections);
+      ::createIndex(StaticStrings::StatisticsRawCollection,
+                    arangodb::Index::TRI_IDX_TYPE_SKIPLIST_INDEX, {"time"},
+                    false, false, collections);
+    } catch (Result res) {
+      return {res};
+    } catch (...) {
+      return {TRI_ERROR_INTERNAL};
+    }
+  }
+  return {TRI_ERROR_NO_ERROR};
+}
+
+Result createSystemCollectionsIndices(TRI_vocbase_t& vocbase,
+                                      std::vector<std::shared_ptr<LogicalCollection>>& collections) {
+  try {
+    if (vocbase.isSystem()) {
+      ::createIndex(StaticStrings::UsersCollection, arangodb::Index::TRI_IDX_TYPE_HASH_INDEX,
+                    {"user"}, true, true, collections);
+
+      Result res = ::createSystemStatisticsIndices(vocbase, collections);
+      if (res.fail()) {
+        return {res};
+      }
+    }
+
+    upgradeGeoIndexes(vocbase);
+
+    ::createIndex(StaticStrings::AppsCollection, arangodb::Index::TRI_IDX_TYPE_HASH_INDEX,
+                  {"mount"}, true, true, collections);
+    ::createIndex(StaticStrings::JobsCollection, arangodb::Index::TRI_IDX_TYPE_SKIPLIST_INDEX,
+                  {"queue", "status", "delayUntil"}, false, false, collections);
+    ::createIndex(StaticStrings::JobsCollection, arangodb::Index::TRI_IDX_TYPE_SKIPLIST_INDEX,
+                  {"status", "queue", "delayUntil"}, false, false, collections);
+  } catch (Result res) {
+    return {res};
+  } catch (...) {
+    return {TRI_ERROR_INTERNAL};
+  }
+  return {TRI_ERROR_NO_ERROR};
 }
 
-bool UpgradeTasks::setupGraphs(TRI_vocbase_t& vocbase,
-                               arangodb::velocypack::Slice const& slice) {
-  ::createSystemCollection(vocbase, "_graphs"); // throws on error
+}  // namespace
+
+bool UpgradeTasks::createSystemCollectionsAndIndices(TRI_vocbase_t& vocbase,
+                                                     arangodb::velocypack::Slice const& slice) {
+  Result res;
+
+  // This vector should after the call to ::createSystemCollections contain
+  // a LogicalCollection for *every* (required) system collection.
+  std::vector<std::shared_ptr<LogicalCollection>> presentSystemCollections;
+  res = ::createSystemCollections(vocbase, presentSystemCollections);
+
+  // TODO: Maybe check or assert that all collections are present (i.e. were
+  //       present or created), raise an error if not?
+
+  if (res.fail()) {
+    LOG_TOPIC("e32fi", ERR, Logger::MAINTENANCE)
```

Comment:
Is this log here correct?
Maintenance is only on DBServers, this Code is executed on all types of servers.
If the message is moved from somewhere else to here, remember to modify the logid